### PR TITLE
Add git blame error reporting with notification

### PR DESCRIPTION
**Pull request by mmtftr**
Screenshot 2024-04-11 at 13 13 44

Release Notes:

- Added a notification to show git blame errors.

Caveats:
- ~git blame now executes in foreground
executor  (required since the Fut is !Send)~

TODOs:
- After a failed toggle, the app thinks the blame
is shown. This means toggling again will do nothing
instead of retrying. (Caused by editor.show_git_blame
being set to true before the git blame is generated)
- ~(Maybe) Trim error?~ Done

### CODE REVIEW

File: crates/project/src/project.rs [hunk 1]
```diff
@@ -7708,15 +7709,29 @@ impl Project {
...
+            cx.to_async().spawn(|mut cx| {
+                let this = this.clone();
+                async move {
+                    let (repo, relative_path, content) = blame_params?;
+                    let lock = repo.lock();
+                    let blame_res = lock
+                        .blame(&relative_path, content)
+                        .with_context(|| format!("Failed to git blame {relative_path:?}"));
+
+                    if let Err(ref err) = blame_res {
```
Comment (osiewicz) [inline crates/project/src/project.rs:7721]: I'm not @mrnugget , but I guess forcing this to be on a foreground thread is a pretty big deal; wouldn't it work if we emitted a notification on error in the caller of `blame_buffer`?

Comment (mmtftr) [reply crates/project/src/project.rs]: yeah, good point. did exactly that.

File: crates/editor/src/git/blame.rs [hunk 2]
```diff
@@ -253,10 +253,10 @@ impl GitBlame {
     fn generate(&mut self, cx: &mut ModelContext<Self>) {
         let buffer_edits = self.buffer.update(cx, |buffer, _| buffer.subscribe());
         let snapshot = self.buffer.read(cx).snapshot();
-        let blame = self.project.read(cx).blame_buffer(&self.buffer, None, cx);
+        let blame = Project::blame_buffer(&self.project, &self.buffer, None, cx);
```
Comment (mrnugget) [inline crates/editor/src/git/blame.rs:256]: I'm not sure I'm following — why do we need this change?

Comment (mmtftr) [reply crates/editor/src/git/blame.rs:256]: I initially needed it to get access to the Model and ModelContext objects but now I don't. I'll revert that part and push another commit

**Comment by mrnugget** [comment]
More a higher-level question: will we display errors to the user that are actionable? In your screenshot, for example, is that something that the user can fix, or is that a bug on our side?

If it's not actionable for the user, should we instead just `.log_err()`?

**Comment by mmtftr** [comment]
In my example, it's actionable. My git-blame-ignore-revs file was missing which caused an error with git blame. I could not figure this out until I manually ran zed from the console to see the error.

In the case that git blame fails for whatever reason, the user should see that it failed in my opinion. Without the notification, the experience is that the user toggles the git blame but nothing happens.

**Comment by mrnugget** [comment]
> In my example, it's actionable. My git-blame-ignore-revs file was missing which caused an error with git blame. I could not figure this out until I manually ran zed from the console to see the error.

Okay, that makes sense. Yeah, I think if we can simplify the code more so that it only shows the error, that makes sense.

**Comment by mmtftr** [comment]
> If it's not actionable for the user, should we instead just `.log_err()`?

I wanted to `log_err()` but for some reason that returns an Option, (I think it should be like inspect_err and not change the type).

**Comment by mmtftr** [comment]
Added test coverage. Simplified the error context. The final error looks like this when the process fails: 
Screenshot 2024-04-11 at 20 30 23

**Comment by mrnugget** [comment]
Thanks. I hopped on the branch and simplified the code a little bit. Will merge once CI is through.

**Pull request merged**